src/lib/tuning/index/base_index_evaluator.cpp

```diff
@@ -0,0 +1,222 @@
+#include "base_index_evaluator.hpp"
+
+#include <algorithm>
+#include <list>
+#include <memory>
+#include <string>
+
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/stored_table_node.hpp"
+#include "operators/get_table.hpp"
+#include "operators/table_scan.hpp"
+#include "operators/validate.hpp"
+#include "optimizer/column_statistics.hpp"
+#include "optimizer/table_statistics.hpp"
+#include "sql/gdfs_cache.hpp"
+#include "sql/sql_query_cache.hpp"
+#include "storage/index/base_index.hpp"
+#include "storage/storage_manager.hpp"
+#include "types.hpp"
+#include "utils/logging.hpp"
+
+namespace opossum {
+
+BaseIndexEvaluator::BaseIndexEvaluator() {}
+
+void BaseIndexEvaluator::evaluate(std::vector<std::shared_ptr<TuningChoice>>& choices) {
+  // Allow concrete implementation to initialize
+  _setup();
+
+  // Scan query cache for indexable table column accesses
+  _inspect_query_cache();
+
+  // Aggregate column accesses to set of new columns to index
+  _aggregate_access_records();
+
+  // Fill index_evaluations vector
+  _choices.clear();
+  _add_existing_indexes();
+  _add_new_indexes();
+
+  // Evaluate
+  for (auto& index_choice : _choices) {
+    if (index_choice.exists) {
+      index_choice.memory_cost = static_cast<float>(_existing_memory_cost(index_choice));
+    } else {
+      index_choice.type = _propose_index_type(index_choice);
+      index_choice.memory_cost = static_cast<float>(_predict_memory_cost(index_choice));
+    }
+    index_choice.saved_work = _calculate_saved_work(index_choice);
+
+    // Transfer results to choices vector
+    choices.push_back(std::make_shared<IndexChoice>(index_choice));
+  }
+}
+
+void BaseIndexEvaluator::_setup() {}
+
+void BaseIndexEvaluator::_process_access_record(const BaseIndexEvaluator::AccessRecord&) {}
+
+uintptr_t BaseIndexEvaluator::_existing_memory_cost(const IndexChoice& index_choice) const {
+  const auto table = StorageManager::get().get_table(index_choice.column_ref.table_name);
+  uintptr_t memory_cost = 0u;
+  for (ChunkID chunk_id = ChunkID{0}; chunk_id < table->chunk_count(); ++chunk_id) {
+    const auto chunk = table->get_chunk(chunk_id);
+    const auto index = chunk->get_index(index_choice.type, index_choice.column_ref.column_ids);
+    if (index) {
+      memory_cost += index->memory_consumption();
+    }
+  }
+  return memory_cost;
+}
+
+void BaseIndexEvaluator::_inspect_query_cache() {
+  _access_records.clear();
+
+  /*
+   * ToDo(anybody): The cache interface could be improved by introducing values() method in
```

Comment:
Please add a link to this part to the issue that was opened for this so that people can find the piece of code it refers to

Reply:
done